compiler/sem/tailcall_elim.nim

```diff
@@ -0,0 +1,525 @@
+## Implements both the semantic analysis part of tail-call elimination and
+## the actual elimination. Refer to this `document <tailcallelim.html>`_ for
+## a high-level overview of how tail-call elimination works.
+##
+## The `apply` and `trampoline` helper procedures are created during semantic
+## analysis already, as it potentially requires instantiating generic types.
+## The synthesized procedures are attached to their originating ones via the
+## misc slot.
+
+import
+  compiler/ast/[
+    ast,
+    ast_query,
+    ast_types,
+    ast_idgen,
+    idents,
+    lineinfos,
+    trees,
+    types
+  ],
+  compiler/modules/[
+    magicsys,
+    modulegraphs
+  ],
+  compiler/sem/[
+    lowerings,
+    semdata,
+    semtypinst
+  ],
+  compiler/front/[
+    msgs,
+    options
+  ]
+
+from compiler/ast/report_enums import ReportKind
+from compiler/ast/reports_sem import SemReport, reportStr
+
+type
+  Mode = enum
+    emDefer
+    emTry
+    emExpr
+    emLast
+
+proc error(config: ConfigRef, info: TLineInfo, msg: string) =
+  # TODO: use dedicated reports
+  config.localReport(info, reportStr(rsemUserError, msg))
+
+# -------------- analysis --------------
+
+proc checkArg(g: ModuleGraph, owner: PSym, n: PNode, i: int, formal: PType) =
+  ## Analyses the `i`-th argument `n` for an call to a procedure with type
+  ## `formal`. Reports an error for arguments not adhering to the .musttail
+  ## call rules.
+  proc isValid(owner, s: PSym): bool =
+    (sfGlobal in s.flags) or (s.kind == skParam and s.owner == owner)
+
+  case formal[i].kind
+  of tySink:
+    discard "all expressions are admitted"
+  of tyVar:
+    let root = getRoot(n)
+    # globals and our own parameters are okay, everything else is not
+    if not isValid(owner, root):
+      g.config.error(n.info, "argument must borrow from a parameter or global")
+  else:
+    if isPassByRef(g.config, formal.n[i].sym, formal[0]) or
+        hasDestructor(formal[i]):
+      # pass-by-value arguments with custom copy behaviour cannot safely be
+      # shallow-copied and thus must be borrowed too
+      let root = getRoot(n)
+      if root.isNil:
+        g.config.error(n.info, "argument must be lvalue expression")
+      elif not isValid(owner, root):
+        g.config.error(n.info, "argument must borrow from a parameter or global")
+    else:
+      discard "acts like a sink parameter"
+
+proc supportsSiblingCalls(s: PSym): bool =
+  # sibling-calling (necessary for tail-call elimination) requires modifying
+  # a caller's signature, therefore, as a rule of thumb: every routine whose
+  # signature cannot easily and transparently be changed to something else
+  # cannot use tail calls
+  if s.typ == nil:
+    echo s
```

Comment:
Leftover debug code?

Reply:
Yep, thanks, I'd have likely missed that.